go/arrow/csv/writer.go

```diff
@@ -0,0 +1,147 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package csv reads CSV files and presents the extracted data as records.
+package csv
+
+import (
+	"encoding/csv"
+	"fmt"
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"io"
+)
+
+// WriterOption configures a CSV writer.
+type WriterOption func(*Writer)
+
+// WithComma specifies the fields separation character used while writing CSV files.
+func WithFieldSeparator(c rune) WriterOption {
+	return func(w *Writer) {
+		w.w.Comma = c
+	}
+}
+
+// UseCRLF specifies the line terminator used while writing CSV files.
+func WithLineTerminator(b bool) WriterOption {
+	return func(w *Writer) {
+		w.w.UseCRLF = b
+	}
+}
+
+// Reader wraps encoding/csv.Reader and creates array.Records from a schema.
+type Writer struct {
+	w      *csv.Writer
+	schema *arrow.Schema
+}
+
+// NewWriter returns a writer that write array.Records to the CSV file
+// with the given schema.
+//
+// NewWriter panics if the given schema contains fields that have types that are not
+// primitive types.
+func NewWriter(r io.Writer, schema *arrow.Schema, opts ...WriterOption) *Writer {
+	validate(schema)
+
+	rr := &Writer{w: csv.NewWriter(r), schema: schema}
+	for _, opt := range opts {
+		opt(rr)
+	}
+
+	return rr
+}
+
+func (w *Writer) Schema() *arrow.Schema { return w.schema }
+
+// Write writes a single Record as one row to the CSV file
+func (w *Writer) Write(record array.Record) error {
+	if !record.Schema().Equal(w.schema) {
+		return ErrMismatchFields
+	}
+
+
+	recs := make([][]string, record.NumRows())
+	for i := range recs {
+		recs[i] = make([]string, record.NumCols())
+	}
+
+	for j, col := range record.Columns() {
+		switch w.schema.Field(j).Type.(type) {
+		case *arrow.BooleanType:
+			arr := col.(*array.Boolean)
+			for i := 0; i < arr.Len(); i++ {
+				recs[i][j] = fmt.Sprintf("%v", arr.Value(i))
+			}
+		case *arrow.Int8Type:
+			arr := col.(*array.Int8)
+			for i := 0; i < arr.Len(); i++ {
+				recs[i][j] = fmt.Sprintf("%v", arr.Value(i))
```

Comment:
it might be interesting performance-wise to directly use `strconv.Format{Bool,Float,Int,Uint}` instead of `fmt.Sprintf`.

(this can be done in a later PR, but please write a FIXME or send a JIRA issue.)

Reply:
https://issues.apache.org/jira/browse/ARROW-4735